File: word-crush/syllable_filter.js
const express = require("express");
const fs = require("fs").promises; // For async file operations
const app = express();
const port = 3000;

// Middleware to parse JSON bodies
app.use(express.json());

// Function to remove invalid words
const invalidChars = ["-", ",", "_", "*", "/", "\\"];

const isValidWord = (text) => {
  if (!text) return false;
  return !invalidChars.some((char) => text.includes(char));
};

const removeInvalidWords = (data) => {
  return data.filter((text) => isValidWord(text));
};

// Function to remove duplicate texts
const removeDuplicates = (data) => {
  const uniqueTexts = new Set();
  return data.filter((text) => {
    if (uniqueTexts.has(text)) {
      return false;
    }
    uniqueTexts.add(text);
    return true;
  });
};

// Function to count syllables in Vietnamese words
const countSyllablesVietnamese = (word) => {
  const parts = word.split(" ").filter((part) => part); // Remove empty parts
  return parts.length;
};

// Function to lowercase Vietnamese words
const lowercaseVietnamese = (word) => {
  const parts = word.split(" ");
  const lowercaseParts = parts.map((part) => part.toLowerCase());
  return lowercaseParts.join(" ");
};

// API endpoint to process the input file
app.get("/process-words", async (req, res) => {
  const inputFilePath = "words.txt";
  const outputFilePath = "filtered_output.json";
  let dataList = [];
  let filteredData = [];

  try {
    // Read the input file
    const fileContent = await fs.readFile(inputFilePath, "utf-8");
    const lines = fileContent.split("\n").filter((line) => line.trim());

    // Parse each line as JSON
    for (const line of lines) {
      try {
        const parsedItem = JSON.parse(line);
        dataList.push(parsedItem.text);
      } catch (error) {
        console.error(`Error parsing line: ${line}`);
      }
    }

    // Remove invalid words
    dataList = removeInvalidWords(dataList);

    // Remove duplicate words
    dataList = removeDuplicates(dataList);

    // Filter for 2-syllable words and capitalize
    for (const text of dataList) {
      const syllableCount = countSyllablesVietnamese(text);
      if (syllableCount === 2) {
        const lowercasedText = lowercaseVietnamese(text);
        filteredData.push(lowercasedText);
      }
    }

    // Prepare output data
    const outputData = {
      directoryVNData: filteredData,
    };

    // Save to output file
    await fs.writeFile(
      outputFilePath,
      JSON.stringify(outputData, null, 4),
      "utf-8"
    );
    console.log(`Exported JSON file successfully to: ${outputFilePath}`);

    // Send response
    res.json({
      message: "Processing complete",
      data: outputData,
    });
  } catch (error) {
    if (error.code === "ENOENT") {
      res.status(404).json({ error: `File not found: ${inputFilePath}` });
    } else {
      res.status(500).json({ error: "Internal server error" });
    }
    console.error(error);
  }
});

// Start the server
app.listen(port, () => {
  console.log(`Server running at http://localhost:${port}`);
});

// Endpoint to return just an array of text values
app.get("/text-array", async (req, res) => {
  try {
    const outputFilePath = "filtered_output.json";
    const exportFilePath = "text_array.json";
    const fileContent = await fs.readFile(outputFilePath, "utf-8");
    const data = JSON.parse(fileContent);
    
    const textArray = data.directoryVNData;
    
    // Create the export data
    const exportData = {
      count: textArray.length,
      texts: textArray,
    };
    
    // Write to export file
    await fs.writeFile(
      exportFilePath,
      JSON.stringify(exportData, null, 4),
      "utf-8"
    );
    console.log(`Exported text array JSON file successfully to: ${exportFilePath}`);
    
    // Send response
    res.json({
      message: "Text array exported successfully",
      filePath: exportFilePath,
      count: textArray.length,
      texts: textArray,
    });
  } catch (error) {
    res.status(500).json({ error: "Failed to retrieve text array" });
    console.error(error);
  }
});

File: word-crush/syllable_filter.py
import json

def remove_invalid_words(data):
    """
    Remove words containing invalid characters from the data list.
    
    Args:
        data: List of dictionaries containing 'text' keys
        
    Returns:
        List of dictionaries with filtered text entries
    """
    invalid_chars = ['-', ',', '_', '*', '/', '\\']
    
    def is_valid_word(text):
        if not text:
            return False
        return not any(char in text for char in invalid_chars)
    
    return [item for item in data if is_valid_word(item.get('text', ''))]

def count_syllables_vietnamese(word):
    parts = word.split()
    return len(parts)

def capitalize_vietnamese(word):
    """
    Capitalize each part of a Vietnamese word.
    
    Args:
        word: String containing Vietnamese word with space-separated parts
        
    Returns:
        String with each part capitalized
    
    Example:
        "nguyen van" -> "Nguyen Van"
    """
    parts = word.split()
    capitalized_parts = [part.capitalize() for part in parts]
    return " ".join(capitalized_parts)  # Use space instead of hyphen

input_file_path = 'words.txt'
output_file_path = 'filtered_output.json'
data_list = []
filtered_data = []

try:
    with open(input_file_path, 'r', encoding='utf-8') as file:
        lines = file.readlines()
except FileNotFoundError:
    print(f"File not found: {input_file_path}")
    exit()

for line in lines:
    try:
        item = json.loads(line.strip())
        data_list.append(item)
    except json.JSONDecodeError:
        print(f"Error: {line}")
        continue


data_list = remove_invalid_words(data_list)

for item in data_list:
    text = item['text']
    syllable_count = count_syllables_vietnamese(text)
    if syllable_count == 2:
        capitalized_text = capitalize_vietnamese(text)
        new_item = {
            "text": capitalized_text,
            "picked": "false"
        }
        filtered_data.append(new_item)

output_data = {
    "directoryVNData": filtered_data
}

# Export json file
with open(output_file_path, 'w', encoding='utf-8') as json_file:
    json.dump(output_data, json_file, ensure_ascii=False, indent=4)

print(f"Export json file succesfully in: {output_file_path}")